v1/mutate/mutate.go

```diff
@@ -241,20 +240,35 @@ func validate(adds []Addendum) error {
 	return nil
 }
 
-// Extract takes an image, flattens its filesystem and returns
-// an io.Reader to read the contents of the final fs
+// Extract takes an image and returns an io.ReadCloser containing the image's
+// flattened filesystem.
+//
 // Adapted from https://github.com/google/containerregistry/blob/master/client/v2_2/docker_image_.py#L731
-func Extract(img v1.Image) (io.Reader, error) {
-	var b bytes.Buffer
-	writer := bufio.NewWriter(&b)
-	tarWriter := tar.NewWriter(writer)
+func Extract(img v1.Image) io.ReadCloser {
+	pr, pw := io.Pipe()
+
+	go func() {
+		// Close the writer with any errors encountered during
+		// extraction. These errors will be returned by the reader end
+		// on subsequent reads. If err == nil, the reader will return
+		// EOF.
+		pw.CloseWithError(extract(img, pw))
+	}()
+
+	return pr
+}
+
+func extract(img v1.Image, w io.Writer) error {
+	tarWriter := tar.NewWriter(w)
 	defer tarWriter.Close()
 
 	fileMap := map[string]bool{}
 
 	layers, err := img.Layers()
 	if err != nil {
-		return nil, fmt.Errorf("Error retrieving image layers: %v", err)
+		log.Printf("RETURNING ERROR: %v", err) // TODO remove
+
+		return fmt.Errorf("Error retrieving image layers: %v", err)
```

Comment:
Since you're in here... I think idiomatic go errors are lowercase.

(and I think "Error" is probably redundant to include in the error message)